utilities/cache_dump_load_impl.cc

```diff
@@ -0,0 +1,365 @@
+  if (!io_s.ok()) {
+    return io_s;
+  }
+
+  while (io_s.ok() && dump_unit.type != CacheDumpUnitType::kFooter) {
+    dump_unit.reset();
+    data.clear();
+    io_s = ReadCacheBlock(&data, &dump_unit);
+    if (!io_s.ok()) {
+      break;
+    }
+    BlockContents raw_block_contents(
+        Slice((char*)dump_unit.value, dump_unit.value_len));
+    Cache::CacheItemHelper* helper = nullptr;
+    Statistics* statistics = nullptr;
+    void* block = nullptr;
+    switch (dump_unit.type) {
+      case CacheDumpUnitType::kDeprecatedFilterBlock: {
+        helper = BlocklikeTraits<BlockContents>::GetCacheItemHelper(
+            BlockType::kFilter);
+        std::unique_ptr<BlockContents> block_holder;
+        block_holder.reset(BlocklikeTraits<BlockContents>::Create(
+            std::move(raw_block_contents), 0, statistics, false,
+            toptions_.filter_policy.get()));
+        block = (void*)block_holder.release();
+        break;
+      }
+      case CacheDumpUnitType::kFilter: {
+        helper = BlocklikeTraits<ParsedFullFilterBlock>::GetCacheItemHelper(
+            BlockType::kFilter);
+        std::unique_ptr<ParsedFullFilterBlock> block_holder;
+        block_holder.reset(BlocklikeTraits<ParsedFullFilterBlock>::Create(
+            std::move(raw_block_contents), toptions_.read_amp_bytes_per_bit,
+            statistics, false, toptions_.filter_policy.get()));
+        block = (void*)block_holder.release();
+        break;
+      }
+      case CacheDumpUnitType::kData: {
+        helper = BlocklikeTraits<Block>::GetCacheItemHelper(BlockType::kData);
+        std::unique_ptr<Block> block_holder;
+        block_holder.reset(BlocklikeTraits<Block>::Create(
+            std::move(raw_block_contents), toptions_.read_amp_bytes_per_bit,
+            statistics, false, toptions_.filter_policy.get()));
+        block = (void*)block_holder.release();
+        break;
+      }
+      case CacheDumpUnitType::kIndex: {
+        helper = BlocklikeTraits<Block>::GetCacheItemHelper(BlockType::kIndex);
+        std::unique_ptr<Block> block_holder;
+        block_holder.reset(BlocklikeTraits<Block>::Create(
+            std::move(raw_block_contents), 0, statistics, false,
+            toptions_.filter_policy.get()));
+        block = (void*)block_holder.release();
+        break;
+      }
+      case CacheDumpUnitType::kFilterMetaBlock: {
+        helper = BlocklikeTraits<Block>::GetCacheItemHelper(BlockType::kFilter);
+        std::unique_ptr<Block> block_holder;
+        block_holder.reset(BlocklikeTraits<Block>::Create(
+            std::move(raw_block_contents), toptions_.read_amp_bytes_per_bit,
+            statistics, false, toptions_.filter_policy.get()));
+        block = (void*)block_holder.release();
+        break;
+      }
+      case CacheDumpUnitType::kFooter:
+        break;
+      default:
+        continue;
+    }
+    if (helper != nullptr) {
+      secondary_cache_->Insert(dump_unit.key, block, helper);
```

Comment:
Since we have the object, it can be inserted into the block cache instead of the secondary cache. Have you considered that? This is probably fine too, but just wondering about the pros and cons.

Reply:
Last time you mentioned that we may not have the run time parameter: "read_amp_bytes_per_bit" and "Statistics* statistics" when call the BlocklikeTraits<Block>Create API. It may cause some issue when insert it to the block cache directly without having the DB?

Reply:
That's a good point. Since the cache load is not going through the DB object, it doesn't have access to those parameters. In any case, we can't assume that the DB is opened at the time of loading.

Reply:
Yes, for now, I think it is ok to just support the SecondryCache, which can load blocks at any time and there is no limit. In the future, we can add the new interface that support loading to block cache (which may need to take the DB pointer). So, I may need to rename the function in CacheDumpedLoader from Prepare/Run to RestoreCacheEntriesToSecondaryCache?

Reply:
I don't have a strong opinion about it. Maybe you can just document in the comments that the entries are inserted into the secondary cache.